File: SexyBiscuit.Engine/Core/Time.cs
namespace SexyBiscuit.Engine.Core;

/// <summary>
/// Global frame timing. Advanced once per frame by <see cref="SBEngine"/>;
/// every gameplay system reads from here rather than threading a delta around.
/// </summary>
/// <remarks>
/// <see cref="DeltaTime"/> is already scaled by <see cref="TimeScale"/>, so setting
/// <c>Time.TimeScale = 0</c> freezes gameplay while UI and input keep running off
/// <see cref="UnscaledDeltaTime"/>.
/// </remarks>
public static class Time
{
    /// <summary>Scaled seconds since the previous frame. Clamped by <see cref="MaximumDeltaTime"/>.</summary>
    public static float DeltaTime { get; private set; }

    /// <summary>Real seconds since the previous frame, ignoring <see cref="TimeScale"/>.</summary>
    public static float UnscaledDeltaTime { get; private set; }

    /// <summary>Scaled seconds elapsed since engine start.</summary>
    public static float TimeSinceStartup { get; private set; }

    /// <summary>Real seconds elapsed since engine start, ignoring <see cref="TimeScale"/>.</summary>
    public static float RealtimeSinceStartup { get; private set; }

    /// <summary>Length of one fixed-update step in scaled seconds.</summary>
    public static float FixedDeltaTime { get; internal set; } = 1f / 60f;

    /// <summary>Number of frames rendered since engine start.</summary>
    public static long FrameCount { get; private set; }

    /// <summary>
    /// Multiplier applied to <see cref="DeltaTime"/> and <see cref="FixedDeltaTime"/>.
    /// 0 pauses gameplay, 0.5 is slow motion, 2 is double speed. Never negative.
    /// </summary>
    public static float TimeScale
    {
        get => _timeScale;
        set => _timeScale = MathF.Max(0f, value);
    }
    private static float _timeScale = 1f;

    /// <summary>
    /// Upper bound on a single frame's unscaled delta, in seconds. Prevents a hitch or a
    /// breakpoint from producing a delta large enough to tunnel objects through walls.
    /// </summary>
    public static float MaximumDeltaTime { get; set; } = 0.1f;

    /// <summary>Smoothed frames per second, averaged over roughly the last half second.</summary>
    public static float Fps { get; private set; }
    private static float _fpsAccum;
    private static int   _fpsFrames;

    /// <summary>Called once per frame by the engine. Not part of the game-facing API.</summary>
    internal static void Advance(float rawDeltaSeconds)
    {
        UnscaledDeltaTime     = MathF.Min(rawDeltaSeconds, MaximumDeltaTime);
        DeltaTime             = UnscaledDeltaTime * _timeScale;
        RealtimeSinceStartup += UnscaledDeltaTime;
        TimeSinceStartup     += DeltaTime;
        FrameCount++;

        _fpsAccum  += UnscaledDeltaTime;
        _fpsFrames += 1;
        if (_fpsAccum >= 0.5f)
        {
            Fps        = _fpsFrames / _fpsAccum;
            _fpsAccum  = 0f;
            _fpsFrames = 0;
        }
    }

    /// <summary>Resets all counters. Used by tests and when restarting the engine in-process.</summary>
    internal static void Reset()
    {
        DeltaTime = UnscaledDeltaTime = TimeSinceStartup = RealtimeSinceStartup = 0f;
        FrameCount = 0;
        Fps = 0f;
        _fpsAccum = 0f;
        _fpsFrames = 0;
        _timeScale = 1f;
    }
}

File: html5/src/core/Time.js
// -----------------------------------------------------------------------------
// Time — the static clock, mirroring SexyBiscuit.Engine.Core.Time.
//
// `advance` reproduces the C# arithmetic exactly, including the spike clamp and
// the half-second FPS averaging window, so a game that reads Time behaves the
// same under both runtimes.
// -----------------------------------------------------------------------------

let _timeScale = 1;
let _fpsAccum = 0;
let _fpsFrames = 0;

/** Frame timing. Every field is read-only to game code; the host advances it. */
export const Time = {
    /** Seconds since the last frame, already multiplied by {@link Time.timeScale}. */
    deltaTime: 0,

    /** Seconds since the last frame, ignoring time scale. Menus should use this. */
    unscaledDeltaTime: 0,

    /** Scaled seconds accumulated since start-up. */
    timeSinceStartup: 0,

    /** Wall-clock seconds since start-up. */
    realtimeSinceStartup: 0,

    /** The fixed step, scaled. Set from `EngineConfig.fixedTimestep` times the scale. */
    fixedDeltaTime: 1 / 60,

    /** Frames rendered since start-up. */
    frameCount: 0,

    /** Frames per second, averaged over roughly half a second. */
    fps: 0,

    /**
     * The longest frame the engine will admit to, in seconds. A frame that took
     * longer — a breakpoint, a backgrounded tab — is reported as this, so
     * physics does not explode on the frame after a stall.
     */
    maximumDeltaTime: 0.1,

    /** Multiplier on {@link Time.deltaTime}. Clamped at zero; 0 freezes the game. */
    get timeScale() { return _timeScale; },
    set timeScale(value) { _timeScale = Math.max(0, value); },

    /**
     * Advances the clock by one frame. Called by the host, not by game code.
     * @param {number} rawDeltaSeconds Wall-clock seconds since the previous frame.
     */
    advance(rawDeltaSeconds) {
        Time.unscaledDeltaTime = Math.min(rawDeltaSeconds, Time.maximumDeltaTime);
        Time.deltaTime = Time.unscaledDeltaTime * _timeScale;
        Time.realtimeSinceStartup += Time.unscaledDeltaTime;
        Time.timeSinceStartup += Time.deltaTime;
        Time.frameCount++;

        _fpsAccum += Time.unscaledDeltaTime;
        _fpsFrames++;
        if (_fpsAccum >= 0.5) {
            Time.fps = _fpsFrames / _fpsAccum;
            _fpsAccum = 0;
            _fpsFrames = 0;
        }
    },

    /** Returns the clock to its start-up state. */
    reset() {
        Time.deltaTime = 0;
        Time.unscaledDeltaTime = 0;
        Time.timeSinceStartup = 0;
        Time.realtimeSinceStartup = 0;
        Time.frameCount = 0;
        Time.fps = 0;
        _timeScale = 1;
        _fpsAccum = 0;
        _fpsFrames = 0;
    },
};
